### DISCUSSION TITLE
New Version v4.9.11 - Support for logging Map<Id, SObject>, Slack Plugin Improvements

### DISCUSSION THREAD

**Discussion by jongpie** [jongpie/NebulaLogger#463] [Announcements]
#### Apex Logging Improvements

- Added new instance method overload `LogEntryEventBuilder.setRecord(Map<Id, SObject> recordIdToRecord)`
  - For now, I've decided to not add overloads for this in `Logger` directly - I might in a future release, but for now, `setRecord()` can be called by chaining the method to one of Logger's methods, like `Logger.info('some message').setRecord(recordIdToRecord)`
- Updated `System.debug()` calls in `Logger` to start with 'Nebula Logger' instead of just 'Logger' for clarity

#### Admin Improvements

- Updated formula field `LogEntry__c.LoggingLevelWithImage__c` to use a different emoji for INFO (✅) vs lower levels (⚙️ for DEBUG, FINE, FINER, and FINEST)
- Added inline help text to all of the limits-related formula fields on `LogEntry__c` to describe the thresholds used to determine which emojis are shown in the formulas (⛔ ⚠️ ✅)
- Unlocked package-only: Added [`AnimationRule` metadata](https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_animationrule.htm) for `Log__c.Status__c` to show confetti when the status changes 🥳 🎊 

#### Bugfixes
- Fixed #438 in the "Logger Settings" tab that prevented admins from creating new settings records - all thanks goes to @chazwatkins for reporting & fixing this one!
- Fixed #447 by updating `LogEntryEventHandler` to have an extra null check when processing `LogEntryTagRule__mdt` rules - thanks to @ramagrubhubGithub for reporting this 
- Fixed #453 by simplifying `Logger.getUserSettings()` to always use default field values, and making `LoggerSettingsController` responsible for handling prod-specific settings - thanks to @arbokrad for reporting & chatting through this issue with me
  - This keeps the behavior of `Logger.getUserSettings()` consistent in all orgs (instead of prod behaving differently from sandboxes/scratch orgs), while still providing the suggested defaults for a prod org via the `loggerSettings` LWC/custom tab
  - Some default field values on `LoggerSettings__c` have also been updated to provide the most detailed settings by default
  - This bugfix also led to the creation of the new public method `Logger.createSettings()` - this doesn't seem needed in the managed package right now, so `public` seems sufficient (instead of using `global`)

### Slack Plugin Package Changes - `v1.5.2`

- Updated `SlackLoggerPlugin` Apex class to leverage a new field set, `Log__c.SlackNotificationFields`, to support customization of the fields shown in Slack notifications
  - This field set can either be directly modified (although changes may be reverted during upgrades), or by creating & specifying another field set in the custom metadata record `LoggerParameter.SlackLogNotificationFieldSet`
  - The Slack remote site setting has also been renamed from `Slack.remoteSite-meta.xml` to `LoggerSlack.remoteSite-meta.xml` to ensure there's consistency in having some variation of the word "Log" in all metadata. Some orgs unintentionally had a previously-existing remote site setting (also named just "Slack") unintentionally overwritten - this name change should help avoid overwriting existing remote site settings for Slack.

### Installation Info

Moving forward, new release notes will now include this section to provide the `sfdx` installation commands & installation links for any new package versions. Thanks to @MeStrak for pointing out that the CLI commands weren't included anywhere in the docs - including these details in the release notes will make is much easier for everyone to easily install a specific version, using whichever installation method you prefer.

Please note that the included `sfdx` commands use the latest syntax and parameters - these versions of the commands may not work with older versions of `sfdx`, as newer versions of `sfdx` use a simpler syntax, and some command parameters have been renamed.

#### Core Unlocked Package - no namespace

- SFDX CLI: `sfdx package install --apex-compile package --wait 20 --security-type AdminsOnly --package 04t5Y0000015nDxQAI`
- Sandbox: https://test.salesforce.com/packaging/installPackage.apexp?p0=04t5Y0000015nDxQAI
- Production: https://login.salesforce.com/packaging/installPackage.apexp?p0=04t5Y0000015nDxQAI

#### Slack Unlocked Package Plugin - no namespace

- SFDX CLI: `sfdx package install --apex-compile package --wait 20 --security-type AdminsOnly --package 04t5Y0000015nDJQAY`
- Sandbox: https://test.salesforce.com/packaging/installPackage.apexp?p0=04t5Y0000015nDJQAY
- Production: https://login.salesforce.com/packaging/installPackage.apexp?p0=04t5Y0000015nDJQAY

<hr /><em>This discussion was created from the release <a href='https://github.com/jongpie/NebulaLogger/releases/tag/v4.9.11'>Support for logging Map<Id, SObject>, Slack Plugin Improvements</a>.</em>

**Comment by vijaykanugonda** [comment]
Hi Jonathan,

Good Day!!

Recently upgraded to latest version 4.9.11 from 4.6.x and post upgradation logs are not generating. After investigation I seen that in Logger Settings extra option 'Enable Saving' added which needs to be checked then it started generating logs. Please note it in upgrade process.

This needs to enable both on default settings and setup owners .

![image](https://user-images.githubusercontent.com/26404861/227556643-0c9cab97-f3cd-47ba-94f9-e54f6e15b533.png)
![image](https://user-images.githubusercontent.com/26404861/227556702-cbeddb2b-c2b6-4ea5-a345-c1c3d06ce4d6.png)
![image](https://user-images.githubusercontent.com/26404861/227556751-b7ce2f2a-a547-4d29-bb37-d5435064823f.png)

Please add in upgrade process documentation if its not added.

Thanks,
Vijay Kanugonda.

**Reply by jongpie** [reply]
Hi @vijaykanugonda - I try to mention these types of breaking changes in the release notes. In this case, [the release notes for `v4.7.0`](https://github.com/jongpie/NebulaLogger/releases/tag/v4.7.0) state that if you're upgrading, you'll need to make changes to `LoggerSettings__c` records after upgrading. I recommend reviewing the release notes prior to upgrading in the future.

![image](https://user-images.githubusercontent.com/1267157/227752125-eab0e287-4924-451b-9322-52868e5992f8.png)
